### PR TITLE
[AllocBoxToStack] Don't destroy in dead-ends.

### PR DESCRIPTION
**Pull request by nate-chandler**
It is valid to leak a value on paths into dead-end regions. Specifically, it is valid to leak an `alloc_box`.  Thus, "final releases" in dead-end regions may not destroy the box and consequently may not release its contents.  Therefore it's invalid to lower such final releases to `dealloc_stack`s, let alone `destroy_addr`s.  The in-general invalidity of that transformation results in miscompiling whenever a box is leaked and its projected address is used after such final releases.

Fix this by not treating final releases as boundary markers of the `alloc_box` and not lowering them to `destroy_addr`s and `dealloc_stack`s.

rdar://158149082

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocBoxToStack.swift [hunk 1]
```diff
@@ -359,12 +359,25 @@ private func createAllocStack(for allocBox: AllocBoxInst, flags: Flags, _ contex
     stackLocation = asi
   }
 
+  let deadEndBlocks = context.deadEndBlocks
   for destroy in getFinalDestroys(of: allocBox, context) {
     let loc = allocBox.location.asCleanup.withScope(of: destroy.location)
     Builder.insert(after: destroy, location: loc, context) { builder in
-      if  !unboxedType.isTrivial(in: allocBox.parentFunction), !(destroy is DeallocBoxInst) {
+      if deadEndBlocks.isDeadEnd(destroy.parentBlock) {
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocBoxToStack.swift]: This is not correct for the corner case the alloc_box is within an infinite loop, e.g.
```
bb1:
  %1 = alloc_box
  destroy_value %1
  br bb1
```
If we are not inserting the destroy_addr we will leak the value. 

I don't have a good idea how to handle this without replicating half of the ossa-lifetime-completion code. Maybe a simple solution is to just not promote alloc_box in dead-end infinite loops.

Anyway, this problem should be solved once we have complete lifetimes and OSSA throughout the pipeline.
Maybe also worth adding a TODO to remove this check once we have complete lifetimes

File: lib/SILOptimizer/Transforms/AllocBoxToStack.cpp [hunk 2]
```diff
@@ -693,9 +695,19 @@ static bool rewriteAllocBoxAsAllocStack(AllocBoxInst *ABI) {
                          ABI->getBoxType(), ABI->getModule().Types, 0));
   auto Loc = CleanupLocation(ABI->getLoc());
 
+  auto *deb = deba.get(ABI->getFunction());
```
Comment (eeckstein) [inline lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:700]: No need to fix this in the old pass. As long as we can successfully build the stdlib with this and the bootstrapping CI job does not fail, we are fine.

Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:700]: There is no reason not to.  And indeed there is a reason to: it allows the fix to be picked up in 6.2 compilers.

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocBoxToStack.swift (hunk 1 above)
Comment (nate-chandler) [reply SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocBoxToStack.swift]: This is an ancient bug in AllocBoxToStack.  Once we have OSSA throughout the pipeline, we can then have complete lifetimes throughout the pipeline, but this bug would exist with complete lifetimes throughout the OSSA pipeline until OSSA exists throughout the pipeline.

> I don't have a good idea how to handle this without replicating half of the ossa-lifetime-completion code.

We do not have the ability to use OSSACompleteLifetime in non-OSSA, which is where the actually existing issue exists today.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test macos platform

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocBoxToStack.swift (hunk 1 above)
Comment (nate-chandler) [reply SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocBoxToStack.swift]: Fixed this corner case by relying on the invariant that alloc_boxes in dead-end loops are destroyed as discussed.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test macos platform

**Comment by nate-chandler** [comment]
@swift-ci please test macos platform

**Review by eeckstein — APPROVED** [review]
lgtm!
